File: frontend/src/app/services/appointment.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../environment/environment';
import { AppointmentDTO, BookAppointmentRequest } from '../entities/appointment';

@Injectable({ providedIn: 'root' })
export class AppointmentApi {
  private readonly http = inject(HttpClient);
  private readonly base = environment.baseurl;

  /** Free start times ("HH:mm:ss") for a date/service. Never contains patient identities. */
  availableSlots(date: string, serviceId: string): Observable<string[]> {
    const params = new HttpParams().set('date', date).set('serviceId', serviceId);
    return this.http.get<string[]>(`${this.base}appointments/available-slots`, { params });
  }

  book(request: BookAppointmentRequest): Observable<AppointmentDTO> {
    return this.http.post<AppointmentDTO>(`${this.base}appointments`, request);
  }

  ownAppointments(): Observable<AppointmentDTO[]> {
    return this.http.get<AppointmentDTO[]>(`${this.base}account/appointments`);
  }

  cancelOwn(id: string): Observable<void> {
    return this.http.put<void>(`${this.base}account/appointments/${id}/cancel`, {});
  }

  listAll(): Observable<AppointmentDTO[]> {
    return this.http.get<AppointmentDTO[]>(`${this.base}admin/appointments`);
  }

  updateStatus(id: string, status: string): Observable<AppointmentDTO> {
    return this.http.put<AppointmentDTO>(`${this.base}admin/appointments/${id}/status`, { status });
  }
}

File: backend/src/main/java/edu/clinic/service/AppointmentService.java
package edu.clinic.service;

import edu.clinic.dto.AppointmentDTO;
import edu.clinic.dto.AppointmentStatusUpdateRequest;
import edu.clinic.dto.BookAppointmentRequest;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;

public interface AppointmentService {

    /** Free slot start times for a given date and service, in chronological order. */
    List<LocalTime> availableSlots(LocalDate date, String serviceId);

    AppointmentDTO book(String userId, BookAppointmentRequest request);

    List<AppointmentDTO> ownAppointments(String userId);

    void cancelOwn(String userId, String appointmentId);

    List<AppointmentDTO> listAll();

    AppointmentDTO updateStatus(String appointmentId, AppointmentStatusUpdateRequest request);
}
